go/plugins/dotprompt/dotprompt.go

```diff
@@ -321,3 +327,134 @@ func sortSchemaSlices(s *jsonschema.Schema) {
 		sortSchemaSlices(s.Items)
 	}
 }
+
+// WithTools adds tools to the prompt.
+func WithTools(tools ...ai.Tool) PromptOption {
+	return func(p *Prompt) error {
+		if p.Config.Tools != nil {
+			return errors.New("dotprompt.WithTools: cannot set tools more than once")
+		}
+
+		var toolSlice []ai.Tool
+		toolSlice = append(toolSlice, tools...)
+		p.Tools = toolSlice
+		return nil
+	}
+}
+
+// WithDefaultConfig adds default model configuration.
+func WithDefaultConfig(config *ai.GenerationCommonConfig) PromptOption {
+	return func(p *Prompt) error {
+		if p.Config.GenerationConfig != nil {
+			return errors.New("dotprompt.WithDefaultConfig: cannot set config more than once")
+		}
+		p.Config.GenerationConfig = config
+		return nil
+	}
+}
+
+// WithInputType uses the type provided to derive the input schema.
+func WithInputType(input any) PromptOption {
+	return func(p *Prompt) error {
+		if p.Config.InputSchema != nil {
+			return errors.New("dotprompt.WithInputType: cannot set inputType more than once. Set either inputType or inputSchema")
+		}
+
+		p.Config.InputSchema = base.InferJSONSchemaNonReferencing(input)
+		return nil
+	}
+}
+
+// WithInputSchema adds the schema for input to the prompt.
+func WithInputSchema(schema *jsonschema.Schema) PromptOption {
+	return func(p *Prompt) error {
+		if p.Config.InputSchema != nil {
+			return errors.New("dotprompt.WithInputSchema: cannot set inputSchema more than once. Set either inputType or inputSchema")
+		}
+
+		p.Config.InputSchema = schema
+		return nil
+	}
+}
+
+// WithOutputType uses the type provided to derive the output schema.
+func WithOutputType(output any) PromptOption {
+	return func(p *Prompt) error {
+		if p.Config.OutputSchema != nil {
+			return errors.New("dotprompt.WithOutputType: cannot set outputType more than once. Set either outputType or outputSchema")
+		}
+
+		p.Config.OutputSchema = base.InferJSONSchemaNonReferencing(output)
+		p.Config.OutputFormat = ai.OutputFormatJSON
+		return nil
+	}
+}
+
+// WithOutputSchema adds the schema for output to the prompt.
+func WithOutputSchema(schema *jsonschema.Schema) PromptOption {
+	return func(p *Prompt) error {
+		if p.Config.OutputSchema != nil {
+			return errors.New("dotprompt.WithOutputSchema: cannot set outputSchema more than once. Set either outputType or outputSchema")
+		}
+
+		p.Config.OutputSchema = schema
+		p.Config.OutputFormat = ai.OutputFormatJSON
+		return nil
+	}
+}
+
+// WithOutputFormat adds the desired output format to the prompt
+func WithOutputFormat(format ai.OutputFormat) PromptOption {
+	return func(p *Prompt) error {
+		if p.Config.OutputSchema == nil {
+			return errors.New("dotprompt.WithOutputFormat: outputSchema needs to be set, to set outputFormat")
```

Comment:
This isn't quite right. If it's `OutputFormatText`, schema does not need to be set. If it's `OutputFormatJSON`, schema needs to be set. I think we should also be flexible here and if the user has already implicitly set the format by setting the type/schema, we shouldn't error if they also do `WithOutputFormat(ai.OutputFormatJSON) after` since it's valid. Does that make sense?

Reply:
OK, I get it. Will make some changes